refactor(compiler): introduce internal transplanted type

crisbeto opened the pull request:
Adds a new AST for a `TransplantedType` in the compiler which will be used for some upcoming work. A transplanted type is a type node that is defined in one place in the app, but needs to be copied to a different one (e.g. the generated .d.ts). These changes also include updates to the type translator that will rewrite any type references within the type to point to the new context file.

alxhub commented on packages/compiler/src/output/output_ast.ts:
```diff
@@ -92,6 +103,7 @@ export interface TypeVisitor {
   visitExpressionType(type: ExpressionType, context: any): any;
   visitArrayType(type: ArrayType, context: any): any;
   visitMapType(type: MapType, context: any): any;
+  visitTransplantedType(type: TransplantedType<any>, context: any): any;
```
```suggestion
  visitTransplantedType(type: TransplantedType<unknown>, context: any): any;
```

crisbeto committed:
refactor(compiler): introduce internal transplanted type

Adds a new AST for a `TransplantedType` in the compiler which will be used for some upcoming work. A transplanted type is a type node that is defined in one place in the app, but needs to be copied to a different one (e.g. the generated .d.ts). These changes also include updates to the type translator that will rewrite any type references within the type to point to the new context file.

alxhub commented:
@crisbeto I'm retargeting this to patch since it's a refactoring - we try to keep the patch/main branches in sync as much as possible.

alxhub commented:
This PR was merged into the repository by commit 0f40756a8ab5d0dcba3d06e874bf77acfcdd8d1d.
